[flang] Lower omp.workshare to other omp constructs

ivanradanov opened the pull request:
4/4 

There are two points which need some discussion in this PR:

1. We need to make a value computed in a omp.single accessible in all threads of the omp.parallel region. This is achieved by allocating temporary memory outside the omp.parallel and atoring that in the omp.single and then reloading it from all threads. However, from reading the standard I dont think we are guaranteed that the workshare is nested in the omp.parallel so there could be a omp.parallel { func.call @contains_workshare }, then we would not be able to access the omp.parallel. So I think adding support in the runtime to be able to yield a value from a omp.single could be the fix to this.

2. For the temporary allocations above not all types are supported by fir.alloca, so I need to use llvm.alloca and unrealized_cast to be able to allocate a temporary for a fir.ref type. This too can be fixed by introducing yielding from omp.single

1/4 https://github.com/llvm/llvm-project/pull/101443
2/4 https://github.com/llvm/llvm-project/pull/101444
3/4 https://github.com/llvm/llvm-project/pull/101445
4/4 https://github.com/llvm/llvm-project/pull/101446

WIP https://github.com/llvm/llvm-project/pull/104748 adds HLFIR lowering that make use of this pipeline

tblah commented on flang/lib/Optimizer/OpenMP/LowerWorkshare.cpp:
```diff
@@ -0,0 +1,259 @@
+/// In case we do not find it, we fall back to converting the omp.workshare to
+/// omp.single.
+/// To better handle this we should probably enable yielding values out of an
+/// omp.single which will be supported by the omp runtime.
```
Could you use the copyprivate clause to broadcast values from the omp.single to all other threads?

ivanradanov replied:
Done, thank you for the suggestion.

tblah commented on flang/lib/Optimizer/OpenMP/LowerWorkshare.cpp:
```diff
@@ -0,0 +1,259 @@
+      alloc = allocBuilder.create<fir::AllocaOp>(loc, ty);
+      singleBuilder.create<fir::StoreOp>(loc, singleMapping.lookup(v), alloc);
+      reloaded = rootBuilder.create<fir::LoadOp>(loc, ty, alloc);
+    } else {
```
What types are you seeing used which cannot be supported by fir alloca?

ivanradanov replied:
I think building a fir.alloca for a ReferenceType fails and I encountered that somewhere. I will check again to see if I can get an example.

tblah commented on flang/lib/Optimizer/OpenMP/LowerWorkshare.cpp:
```diff
@@ -0,0 +1,259 @@
+    Value alloc, reloaded;
+    if (isSupportedByFirAlloca(ty)) {
+      alloc = allocBuilder.create<fir::AllocaOp>(loc, ty);
+      singleBuilder.create<fir::StoreOp>(loc, singleMapping.lookup(v), alloc);
```
What about more complicated types e.g. an allocatable array? If you only shallow copy the descriptor won't every thread in the parallel region try to free the same allocation?

ivanradanov replied:
I am sorry, this is probably due to my inexperience with flang/fortran, are there any types in flang that get automatically freed on scope exit? Because if I make a shallow copy of an allocatable array, the operation that frees it will be put in a omp.single, thus freeing it only once sincle the free operation would not be `IsSafeToParallelize`.

e.g. 

```
%a = fir.allocmem
use(%a)
fir.freemem %a

->

omp.single{ %a = fir.allocmem}
use(%a)
omp.single{ fir.freemem %a }
```

ivanradanov replied:
I suppose it will be a problem with if we try to do this to a fir.alloca, which goes out of scope, we would need to firstprivate those for the omp.single regions. Thank you for the comment

ivanradanov replied:
I have opted hoist fir.alloca ops to the parallel region and scattering them using omp.single's copyprivate

tblah commented on flang/lib/Optimizer/OpenMP/LowerWorkshare.cpp:
```diff
@@ -0,0 +1,259 @@
+  }
+
+  if (!wsOp.getNowait())
+    rootBuilder.create<omp::BarrierOp>(loc);
```
I think the barrier will be created by the OpenMPIRBuilder when lowering from MLIR to LLVMIR.

ivanradanov replied:
This specific barrier is for the synchronisation at the end of omp.workshare, which is mandated by the standard unless there is a nowait.

tblah commented on flang/include/flang/Optimizer/OpenMP/Passes.td:
```diff
@@ -0,0 +1,18 @@
+
+include "mlir/Pass/PassBase.td"
+
+def LowerWorkshare : Pass<"lower-workshare"> {
```
This pass doesn't have an operation type associated with it and so `pm.addPass` will run it on every operation in the module (and the module itself). I think we can only get workshare operations inside of functions so maybe this should be run on func.func. Maybe for more future proofing you could run it on all top level operations (e.g. `addNestedPassToAllTopLevelOperations` instead of `pm.addPass`).

I think the pass has to be run on the parent of the workshare loop not on the workshare loop operation itself because operations are inserted and removed from that parent.

tblah commented on flang/lib/Optimizer/OpenMP/LowerWorkshare.cpp:
```diff
@@ -0,0 +1,259 @@
+      lowerWorkshare(wsOp);
+    });
+
+    // Do folding
```
Please could you expand on this comment. I don't understand what this part is doing. Are you just trying to canonicalize the parent after lowering the workshare?

ivanradanov replied:
I was trying to get rid of some trivially dead operations which remain after the transformation. I now do that explicitly in the transformation.

tblah reviewed (commented):
Thank you for your work so far. This is a great start.

What is the plan for transforming do loops generated by lowering (e.g. that do not become hlfir.elemental operations and are not generated by hlfir bufferization)?

ivanradanov commented:
> Thank you for your work so far. This is a great start.
> 
> What is the plan for transforming do loops generated by lowering (e.g. that do not become hlfir.elemental operations and are not generated by hlfir bufferization)?

I am looking at [this](https://www.openmp.org/spec-html/5.0/openmpsu39.html) for the standard.

I intend to go through the various constructs that require to be separated into units of work and provide an alternative lowering for them so that they will get parallelized when we lower the workdistribute operation.

To accurately keep track of constructs that need to be parallelized for workdistribute I em debating adding a new loop_nest wrapper for that as discussed [here](https://github.com/llvm/llvm-project/pull/101445#issuecomment-2261837847)

ivanradanov commented on flang/include/flang/Optimizer/CMakeLists.txt:
```diff
@@ -2,3 +2,4 @@ add_subdirectory(CodeGen)
 add_subdirectory(Dialect)
 add_subdirectory(HLFIR)
 add_subdirectory(Transforms)
+add_subdirectory(OpenMP)
```
Thank you, I missed those passes, I guess I will make a separate PR for creating the OpenMP pass directory and moving those there.  I will link that here once done.

ivanradanov replied:
PR for this up here https://github.com/llvm/llvm-project/pull/104732

ivanradanov commented on flang/lib/Optimizer/HLFIR/Transforms/BufferizeHLFIR.cpp:
```diff
@@ -792,7 +793,8 @@ struct ElementalOpConversion
     // fir.elemental region inside the inner loop.
     hlfir::LoopNest loopNest =
-        hlfir::genLoopNest(loc, builder, extents, !elemental.isOrdered());
+        hlfir::genLoopNest(loc, builder, extents, !elemental.isOrdered(),
+                           flangomp::shouldUseWorkshareLowering(elemental));
```
According to the standard:

> For array expressions within each statement, including transformational array intrinsic functions that compute scalar values from arrays:
>    Evaluation of each element of the array expression, including any references to ELEMENTAL functions, is a unit of work.
>    Evaluation of transformational array intrinsic functions may be freely subdivided into any number of units of work.

I was under the impression that ELEMENTAL functions must be PURE, however now I see that they can be marked IMPURE. The standard says to divide these into units of work, and 

> It is unspecified how the units of work are assigned to the threads executing a workshare region. 

So just from reading this it looks like we should lower those to wsloops and the user is responsible for any race conditions.

Or perhaps this part

> An implementation of the workshare construct must insert any synchronization that is required to maintain standard Fortran semantics.

Can be taken to mean that we need to schedule these properly.

Perhaps we can introduce a non-standard clause to the workdistribute to control this?

ivanradanov replied:
I have opted to only parallelize loops which are marked as unordered for now.

ivanradanov commented on flang/lib/Optimizer/OpenMP/LowerWorkshare.cpp:
```diff
@@ -0,0 +1,259 @@
+
+namespace flangomp {
+bool shouldUseWorkshareLowering(Operation *op) {
+  auto workshare = dyn_cast<omp::WorkshareOp>(op->getParentOp());
```
I was misreading the standard that we are only parallelizing statements closely nested in a workdistribute. I will fix this to handle nested control regions as well.

ivanradanov commented:
@kiranchandramohan @tblah 
I think this warrants another look if you have some time. 

I have reiterated a bit and opted to have a omp loop nest wrapper op which signals to the workshare lowering which specific loops need to be parallelized (i.e. converted to wsloop { loop_nest}).

This will allow us to emit this in the frontend if it is needed and be more precise about the exact loops that need to be parallelized.


So the LowerWorksharePass that I have implemented here is tasked with parallelizing the loops nested in workshare_loop_wrapper and both the Fortran->mlir frontend and the hlfir lowering passes would be responsible for emitting the workshare_loop_wrapper ops where appropriate. For that I have started with some of the obvious lowerings in the hlfir bufferizations, but perhaps that can be done gradually and not everything needs to be covered by this PR. Let me know what you think.

ivanradanov committed:
[flang] Lower omp.workshare to other omp constructs

Change to workshare loop wrapper op

Move single op declaration

Schedule pass properly

Correctly handle nested nested loop nests to be parallelized by workshare

Leave comments for shouldUseWorkshareLowering

Use copyprivate to scatter val from omp.single

TODO still need to implement copy function
TODO transitive check for usage outside of omp.single not imiplemented yet

Transitively check for users outisde of single op

TODO need to implement copy func
TODO need to hoist allocas outside of single regions

Add tests

Hoist allocas

More tests

Emit body for copy func

Test the tmp storing logic

Clean up trivially dead ops

Only handle single-block regions for now

Fix tests for custom assembly for loop wrapper

Only run the lower workshare pass if openmp is enabled

Implement some missing functionality

Fix tests

Fix test

Iterate backwards to find all trivially dead ops

Add expalanation comment for createCopyFun

Update test

Emit a proper error message for CFG in workshare

Cleanup tests

Fix todo tests

Fix dst src in copy function

Use omp.single to handle CFG cases

Fix lower workshare tests

Different warning

Fix bug and add better clarification comments

Fix message

Fix tests

Do not emit empty omp.single's

LowerWorkshare tests

pipelines fix

tblah commented on flang/lib/Optimizer/OpenMP/LowerWorkshare.cpp:
```diff
@@ -0,0 +1,399 @@
+         isMemoryEffectFree(op);
+}
+
+static mlir::func::FuncOp createCopyFunc(mlir::Location loc, mlir::Type varType,
```
Could this re-use `createCopyFunc` from `ClauseProcessor.cpp`? It would be good to only implement copyprivate in one place.

If not, please could you document the difference between workshare copyprivate and the implementation of the copyprivate clause.

tblah reviewed (commented):
Thank you for all of the updates!

> So the LowerWorksharePass that I have implemented here is tasked with parallelizing the loops nested in workshare_loop_wrapper and both the Fortran->mlir frontend and the hlfir lowering passes would be responsible for emitting the workshare_loop_wrapper ops where appropriate. For that I have started with some of the obvious lowerings in the hlfir bufferizations, but perhaps that can be done gradually and not everything needs to be covered by this PR. Let me know what you think.

Doing it gradually sounds good to me. When you take this out of draft, please document in the commit message exactly what is and is not supported at this point.

agozillon commented on flang/include/flang/Tools/CLOptions.inc:
```diff
@@ -344,6 +345,7 @@ inline void createHLFIRToFIRPassPipeline(
   pm.addPass(hlfir::createLowerHLFIRIntrinsics());
   pm.addPass(hlfir::createBufferizeHLFIR());
   pm.addPass(hlfir::createConvertHLFIRtoFIR());
+  pm.addPass(flangomp::createLowerWorkshare());
```
Sorry, just seen this ping! The comment is primarily to state that the passes should be ran immediately after lowering from parse tree to IR (HLFIR/FIR/OMP), as they make a lot of changes to convert things into a more final form for the OMP dialect with respect to target. It was previously a lot more important as we had a form of outlining that ripped out target regions from their functions into seperate functions. That's no longer there, but we do still have some passes that modify the IR at this stage to a more finalized form for target, in particular OMPMapInfoFinalization which will generate some new maps for descriptor types, OMPMarkDeclareTarget which will mark functions declare target implicitly, and another that removes functions unnecessary for device. There is also a pass or will be for do concurrent which I believe outlines loops into target regions as well. 

But TL;DR, there's a lot of things going on in those passes that would be preferable to keep happening immediately after lowering from the parse tree so later passes can depend on the information being in the "correct" format, whether or not that "immediate" location has changed to after this HLFIR lowering or remains where it is currently I am unsure of! 

@skatrak @jsjodin may also have some feedback/input to this.

ivanradanov replied:
I opted to keep the rest of the openmp passes as they are and have added a bool argument to control whether to run the lower-workshare pass

tblah commented on flang/lib/Optimizer/OpenMP/LowerWorkshare.cpp:
```diff
@@ -0,0 +1,446 @@
+      bool isLast = i + 1 == regions.size();
+      if (std::holds_alternative<SingleRegion>(opOrSingle)) {
+        OpBuilder singleBuilder(sourceRegion.getContext());
+        Block *singleBlock = new Block();
```
I'm not sure about this. Everywhere else in flang we use `OpBuilder::createBlock`.

I think the only difference would be that builder listeners would be notified, and we don't have any here. But this could be surprising when the code is changed. 

But I can see from the implementation of `createBlock` that this isn't incorrect so it is okay with me if you have a good reason?

ivanradanov replied:
The problem here is that until we construct this omp.single block we do not know how many operands the omp.single will have. So if we would like to construct the omp.single which will be the parent of this block we need to do two passes and the code becomes more complex. (createBlock does not support creating orphan blocks). After we construct this block we know what values we need to broadcast outside of it so we can construct the omp.single parent with the correct operands.

tblah reviewed (commented):
Thank you for all of the updates. This looks good to me in general, although I have a few minor questions.

Before merging please make sure you document in the commit message exactly what is and is not supported.

Please could another reviewer take an in-depth look. This is quite complex.

ivanradanov commented on flang/lib/Optimizer/OpenMP/LowerWorkshare.cpp:
```diff
@@ -0,0 +1,446 @@
+  return false;
+}
+
+/// We clone pure operations in both the parallel and single blocks. this
+/// functions cleans them up if they end up with no uses
+static void cleanupBlock(Block *block) {
+  for (Operation &op : llvm::make_early_inc_range(
+           llvm::make_range(block->rbegin(), block->rend())))
+    if (isOpTriviallyDead(&op))
+      op.erase();
+}
```
It is not necessarily a problem to leave them lying around, but this is a fairly cheap way to retain a somewhat canonical form after the pass.

We can get rid of it if preferable.

ivanradanov commented:
@kiranchandramohan @tblah @skatrak I have a question to people more familiar with Fortran and the entire Flang pipeline - is it possible that we would have CFG (multiple blocks) in the IR generated in the workshare statement at this point in the pipeline (immediately after lowering HLFIR to FIR).

The transformation I implemented can work with CFG but the transformation _inlines_ the region contained in the `omp.workshare` region in its parent op (while transforming it), which means that the parent op would need to support multiple blocks, which is not a given (e.g. `fir.if`). Is there an operation like `scf.execute` which can be used here to inline the contents of the `omp.workshare` or should we not support CFG in this transformation.

tblah commented:
> @kiranchandramohan @tblah @skatrak I have a question to people more familiar with Fortran and the Flang pipeline - is it possible that we would have CFG (multiple blocks) in the IR generated in the `omp.workshare` region at this point in the pipeline (immediately after lowering HLFIR to FIR)?
> 
> The transformation I implemented here can work with CFG but the transformation _inlines_ the region contained in the `omp.workshare` region in its parent op (while transforming it), which means that the parent op would need to support multiple blocks, which is not a given (e.g. `fir.if`). Is there an operation like `scf.execute` which can be used here to inline the contents of the `omp.workshare` or should we not support CFG in this transformation.

Yes we could have multiple blocks in cases where the control flow is not representable using our structured control flow operations. For example using CYCLE, EXIT or STOP statements.

I think as a first pass it would be better not to support CFG, because this has been a frequent source of bugs in OpenMP lowering and will need to be tested carefully.

I think the operation you are looking for might be `scf.execute_region`.

tblah commented:
> I was wondering if there is some op that is like `scf.execute_region` but already used in flang.

Not that I am aware of.

I think adding `scf.execute_region` might be the easiest way to support this. The alternative would be to go back and convert the if statement into CFG before inserting into it. I guess this would warrant another RFC as we don't use scf currently.

ivanradanov commented:
I have added a comment explaining the limitation of not allowing CFG in workshare for now and an appropriate TODO message for that.

Regarding documenting the current supported statements in workshare - this specific commit will still result in the trivial omp.single lowering - for the subsequent PRs which actually enable various types of statements in the workshare construct, I will document what they enable. (https://github.com/llvm/llvm-project/pull/104748)

tblah commented:
Thanks for the updates and good job noticing the bug with unstructured control flow.

> I have added a comment explaining the limitation of not allowing CFG in workshare for now and an appropriate TODO message for that.

My concern with the TODO message is that some code that previously compiled using the lowering of WORKSHARE as SINGLE will now hit this TODO. This is okay with me so long as it is fixed soon (before LLVM 20). Otherwise, could these cases continued to be lowered as SINGLE for now.

ivanradanov commented:
> My concern with the TODO message is that some code that previously compiled using the lowering of WORKSHARE as SINGLE will now hit this TODO. This is okay with me so long as it is fixed soon (before LLVM 20). Otherwise, could these cases continued to be lowered as SINGLE for now.

I have updated it to lower to omp.single and emit a warning in CFG cases.

tblah commented on flang/lib/Optimizer/OpenMP/LowerWorkshare.cpp:
```diff
@@ -96,6 +94,12 @@ bool shouldUseWorkshareLowering(Operation *op) {
 
+  if (parentWorkshare.getRegion().getBlocks().size() != 1) {
+    parentWorkshare->emitWarning(
+        "omp workshare with unstructured control flow currently unsupported.");
```
nit
```suggestion
        "omp workshare with unstructured control flow is currently unsupported.");
```

ivanradanov commented:
I have rebased this on the latest main and also marked the follow up https://github.com/llvm/llvm-project/pull/104748 as ready for review. This follow up PR contains code and tests which are needed to fully check this implementation as well.

I think this stack is currently in a good state to be merged.

The 1/4 https://github.com/llvm/llvm-project/pull/101443 2/4 https://github.com/llvm/llvm-project/pull/101444 3/4 https://github.com/llvm/llvm-project/pull/101445 are already approved and good to go, but 2/4 https://github.com/llvm/llvm-project/pull/101444 must be merged together with this PR because otherwise it will result in compilation failures for `omp workshare`.

Thus, it would be great if this PR can be reviewed as well and we can proceed with merging if it looks good.


(The build failures are only on windows and coming from the main branch and not introduced by this)

Thirumalai-Shaktivel commented:
All the PRs LGTM, this works for my test cases.

tblah commented:
@ivanradanov is there anything blocking these 4 patches?

ivanradanov commented:
@tblah I think they are in a good state - I just need a review on this one - the other ones are approved.

tblah reviewed (approved):
Looks great to me and this passes the testsuites I have available. Thanks again for all of your work on this.
